flow/layers/container_layer.cc

```diff
@@ -10,6 +10,83 @@ namespace flutter {
 
 ContainerLayer::ContainerLayer() {}
 
+void ContainerLayer::AssignOldLayer(std::shared_ptr<ContainerLayer> old_layer) {
+  old_layer->old_layer_ = nullptr;
+  old_layer_ = old_layer;
+}
+
+void ContainerLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  auto old_container = reinterpret_cast<const ContainerLayer*>(old_layer);
+  auto tree = context->BeginSubtree();
+  DiffChildren(context, old_container);
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
+void ContainerLayer::DiffChildren(DiffContext* context,
+                                  const ContainerLayer* old_layer) {
+  if (context->IsSubtreeDirty()) {
+    for (auto& layer : layers_) {
+      layer->Diff(context, nullptr);
+    }
+  } else {
+    assert(old_layer);
+
+    const auto& prev_layers = old_layer->layers_;
+
+    int first_changed, end, last_cur, last_prev;
+    end = layers_.size();
+    first_changed = 0;
+    last_cur = end;
+    last_prev = prev_layers.size();
+    while (first_changed < last_cur && first_changed < last_prev &&
+           layers_[first_changed]->CanDiff(prev_layers[first_changed].get())) {
+      first_changed++;
+    }
+    while (first_changed < last_cur - 1 && first_changed < last_prev - 1 &&
+           layers_[last_cur - 1]->CanDiff(prev_layers[last_prev - 1].get())) {
+      last_cur--;
+      last_prev--;
+    }
+
+    // old layers that don't match
+    for (int i = first_changed; i < last_prev; ++i) {
+      auto layer = prev_layers[i];
+      auto subtree = context->BeginSubtree();
+      context->MarkSubtreeDirty(layer->paint_region());
+    }
+
+    for (int i = 0; i < end; ++i) {
+      if (i < first_changed) {
+        auto layer = layers_[i];
+        auto prev_layer = prev_layers[i];
+        if (layer == prev_layer && !layer->paint_region().has_readback()) {
+          // for retained layers just add their existing paint region, but only
+          // if there is no readback in subtree
+          context->AddPaintRegion(layer->paint_region());
+        } else {
+          layer->Diff(context, prev_layer.get());
```

Comment:
What's the purpose of recursing when a retained layer has readback? Is it so that the readback regions can be added to the context? I want to make sure I understand all of the considerations there.

Reply:
Indeed. I somewhat improved the comment in next iteration: https://github.com/flutter/engine/pull/21409/commits/a091548da47447d27028cd0b342fa48ec6d032d2#diff-fc0e3d96b0f5ffcb2ca12dd80cf73019R79

Layers with readback should be able to register the readback region with DiffContext. There's also a bit of future proofing here: Diff would be the place to determine that backdrop snapshot needs to be discarded, so it's worth recursing into.